src/error-table.rkt

```diff
@@ -1,20 +1,365 @@
 #lang racket
 
-(require "points.rkt" "syntax/types.rkt" "core/localize.rkt" "timeline.rkt"
-         "common.rkt")
+(require racket/flonum racket/set)
+(require "points.rkt" "syntax/types.rkt" "core/localize.rkt" "common.rkt"
+         "ground-truth.rkt")
 
-(provide group-errors)
+(provide actual-errors predicted-errors)
 
-(define (group-errors expr pcontext)  
+(define (actual-errors expr pcontext)
   (match-define (cons subexprs pt-errorss)
     (flip-lists
      (hash->list (car (compute-local-errors (list expr) (*context*))))))
-
+  
   (define pt-worst-subexprs
-    (for/list ([pt-errors (in-list pt-errorss)])
-      (match-define (cons worst-subexpr pt-max-error)
-        (argmax cdr (map cons subexprs pt-errors)))
-      (if (> pt-max-error 2) worst-subexpr #f)))
+    (foldr append
+           '()
+           (for/list ([pt-errors (in-list pt-errorss)]
+                      [(pt _) (in-pcontext pcontext)])
+             (define sub-error (map cons subexprs pt-errors))
+             (define filtered-sub-error (filter (lambda (p) (> (cdr p) 16))
+                                                sub-error))
+             (define mapped-sub-error (map (lambda (p) (cons (car p) pt))
+                                           filtered-sub-error))
+             (if (empty? mapped-sub-error)
+                 (list (cons #f pt))
+                 mapped-sub-error))))
+
+  (for/hash ([group (in-list (group-by car pt-worst-subexprs))])
+    (let ([key (caar group)])
+      (values key (map cdr group)))))
+
+(define is-inexact? list?)
+(define (same-sign? a b)
+  (or (and (positive? a) (positive? b))
+      (and (negative? a) (negative? b))))
+ 
+(define (predicted-errors expr ctx pctx)
+  (define subexprs
+    (all-subexpressions-rev expr (context-repr ctx)))
+  (define subexprs-list (map car subexprs))
+ 
+  (define ctx-list
+    (for/list ([subexpr (in-list subexprs)])
+      (struct-copy context ctx [repr (cdr subexpr)])))
+ 
+  (define subexprs-fn (eval-progs-real subexprs-list ctx-list))
+ 
+  (define error-count-hash
+    (make-hash (map (lambda (x) (cons x '())) (cons #f subexprs-list))))
```

Comment:
Easier to use hash-ref with a default value rather than prepopulate the hash, I think.

Reply:
So, in `main-loop.rkt` I am relying on the fact that the predicted table has all the subexpressions. Say some subexpr we predict to have no error, but the ground truth has error. If I was not looping over a pre-populated table, I wont be able to access this subexpr (because it was never entered into the table), and so I wont access this subexpr in the ground truth as well, and so I won't get correct error counts.